dbms/src/Storages/DeltaMerge/Remote/RNWorkerFetchPages.cpp

```diff
@@ -0,0 +1,250 @@
+    // the moment of calling `occupySpace`.
+    auto page_cache = seg_task->meta.dm_context->db_context.getSharedContextDisagg()->rn_page_cache;
+
+    Stopwatch w_occupy;
+    auto occupy_result = page_cache->occupySpace(cf_tiny_oids, seg_task->meta.delta_tinycf_page_sizes);
+    GET_METRIC(tiflash_disaggregated_breakdown_duration_seconds, type_cache_occupy).Observe(w_occupy.elapsedSeconds());
+
+    return occupy_result;
+}
+
+std::shared_ptr<disaggregated::FetchDisaggPagesRequest> buildFetchPagesRequest(
+    const RNReadSegmentTaskPtr & seg_task,
+    const std::vector<PageOID> & pages_not_in_cache)
+{
+    auto req = std::make_shared<disaggregated::FetchDisaggPagesRequest>();
+    auto meta = seg_task->meta.snapshot_id.toMeta();
+    // The keyspace_id here is not vital, as we locate the table and segment by given
+    // snapshot_id. But it could be helpful for debugging.
+    auto keyspace_id = seg_task->meta.keyspace_id;
+    meta.set_keyspace_id(keyspace_id);
+    meta.set_api_version(keyspace_id == NullspaceID ? kvrpcpb::APIVersion::V1 : kvrpcpb::APIVersion::V2);
+    *req->mutable_snapshot_id() = meta;
+    req->set_table_id(seg_task->meta.physical_table_id);
+    req->set_segment_id(seg_task->meta.segment_id);
+
+    for (auto page_id : pages_not_in_cache)
+        req->add_page_ids(page_id.page_id);
+
+    return req;
+}
+
+RNReadSegmentTaskPtr RNWorkerFetchPages::doWork(const RNReadSegmentTaskPtr & seg_task)
+{
+    auto occupy_result = blockingOccupySpaceForTask(seg_task);
+    auto req = buildFetchPagesRequest(seg_task, occupy_result.pages_not_in_cache);
+    {
+        auto cftiny_total = seg_task->meta.delta_tinycf_page_ids.size();
+        auto cftiny_fetch = occupy_result.pages_not_in_cache.size();
+        LOG_DEBUG(
+            log,
+            "Ready to fetch pages, seg_task={} page_hit_rate={} pages_not_in_cache={}",
+            seg_task->info(),
+            cftiny_total == 0 ? "N/A" : fmt::format("{:.2f}%", 100.0 - 100.0 * cftiny_fetch / cftiny_total),
+            occupy_result.pages_not_in_cache);
+        GET_METRIC(tiflash_disaggregated_details, type_cftiny_read).Increment(cftiny_total);
+        GET_METRIC(tiflash_disaggregated_details, type_cftiny_fetch).Increment(cftiny_fetch);
+    }
+
+    const size_t max_retry_times = 3;
+    std::exception_ptr first_exception;
+
+    // TODO: Maybe don't need to re-fetch all pages when retry.
+    for (size_t i = 0; i < max_retry_times; ++i)
+    {
+        try
+        {
+            doFetchPages(seg_task, req);
+            seg_task->initColumnFileDataProvider(occupy_result.pages_guard);
+
+            // We finished fetch all pages for this seg task, just return it for downstream
+            // workers. If we have met any errors, page guard will not be persisted.
+            return seg_task;
+        }
+        catch (const pingcap::Exception & e)
+        {
+            if (!first_exception)
+                first_exception = std::current_exception();
```

Comment:
Why do we keep the first exception but not the latest exception?

Reply:
Good idea, we should keep the last.